### PR TITLE
Merge changes from dd-trace-java 0.56.0

### PR DESCRIPTION
**Pull request by trask**
https://github.com/DataDog/dd-trace-java/releases/tag/v0.55.0

There are a couple of file format commits at the end, so you may want to filter on the commits when reviewing to avoid those (I just discovered how to do this recently):

image

### CODE REVIEW

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java [hunk 1]
```diff
@@ -16,53 +16,80 @@
...
 @Slf4j
 public class InternalJarURLHandler extends URLStreamHandler {
-  private final Map<String, JarEntry> filenameToEntry = new HashMap<>();
-  private JarFile bootstrapJarFile;
+
+  private static final WeakReference<Pair<String, JarEntry>> NULL = new WeakReference<>(null);
+
+  private final String name;
+  private final FileNotInInternalJar notFound;
+  private final Set<String> packages = new HashSet<>();
```
Comment (anuraaga) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java:44]: Realize it's the same as the previous pattern so optional update, we should probably set this to an `unmodifaibleSet` in the constructor instead of passing a mutable object around?

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java [hunk 2]
```diff
@@ -75,12 +102,32 @@ protected URLConnection openConnection(final URL url) throws IOException {
...
       return new InternalJarURLConnection(url, new ByteArrayInputStream(new byte[0]));
     }
-    final JarEntry entry = filenameToEntry.get(filename);
-    if (null != entry) {
-      return new InternalJarURLConnection(url, bootstrapJarFile.getInputStream(entry));
+    // believe it or not, we're going to get called twice for this,
+    // and the key will be a new object each time.
+    Pair<String, JarEntry> pair = cache.get();
+    if (null == pair || !filename.equals(pair.getLeft())) {
+      StringBuilder sb = JAR_ENTRY_QUERY.get();
```
Comment (anuraaga) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java:109]: Just curious, is it possible for this class to be accessed from multiple threads? I wonder how many, if there are a lot of copies of this buffer, it's unfortunate they all get stuck in memory after classloading is done sort of lean towards not using a threadlocal for it, string seems pretty small.

File: benchmark-integration/jetty-perftest/jetty-perftest.gradle [hunk 3]
```diff
@@ -10,6 +10,6 @@ jar {
   manifest {
     attributes(
       "Main-Class": "io.opentelemetry.perftest.jetty.JettyPerftest"
-    )
+      )
```
Comment (anuraaga) [inline benchmark-integration/jetty-perftest/jetty-perftest.gradle:13]: Intended?

File: gradle.properties [hunk 4]
```diff
@@ -1 +1,4 @@
+org.gradle.parallel=true
```
Comment (anuraaga) [inline gradle.properties:1]: Yay

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java [hunk 5]
```diff
@@ -75,12 +102,32 @@ protected URLConnection openConnection(final URL url) throws IOException {
...
+      sb.append(this.name).append(filename);
+      if (filename.endsWith(".class")) {
+        sb.append("data");
+      }
+      String classFileName = sb.toString();
+      sb.setLength(0);
+      JarEntry entry = bootstrapJarFile.getJarEntry(classFileName);
+      if (null != entry) {
+        pair = Pair.of(filename, entry);
+        // this mechanism intentionally does not ensure visibility of this write, because it doesn't
```
Comment (anuraaga) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java:119]: This is because this is a performance optimization, not for correctness, I'm presuming

Comment (richardstartin) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java:119]: 1. Reference assignment is atomic so this cannot lead to inconsistency. 
2. If some other thread doesn't see the write, they just see another instance or null, and may or may not do extra work.

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java (hunk 1 above)
Comment (richardstartin) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java:44]: This is only used within a very narrow context where (in the Datadog codebase, at least) we will not mutate the set once it's built, making the unmodifiable wrapper a pointless cost. If we wanted to present a view of this set to untrusted callers, we would have wrapped it.

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java (hunk 5 above)
Comment (anuraaga) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java:119]: > may or may not do extra work.

So the answer is yes :) I think the comment can be more self-explantory by talking about the motivation instead of the detail

`// This is a performance optimization, so in the rare event this write is not visible to another thread, it just means the same work is recomputed but does not affect consistency.`

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java (hunk 1 above)
Comment (richardstartin) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java:44]: Note that in this fork, you're not actually using the set at all, so I recommend just removing it and saving a bit of space.

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java [hunk 6]
```diff
@@ -16,53 +16,80 @@
...
 
 @Slf4j
 public class InternalJarURLHandler extends URLStreamHandler {
-  private final Map<String, JarEntry> filenameToEntry = new HashMap<>();
-  private JarFile bootstrapJarFile;
+
+  private static final WeakReference<Pair<String, JarEntry>> NULL = new WeakReference<>(null);
+
+  private final String name;
+  private final FileNotInInternalJar notFound;
```
Comment (iNikem) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java:43]: If I understand correctly, all these changes to this class are performance optimisations, no? Is there any benchmark or, even better, a real use-case, which demonstrate the benefit of this?

File: agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    "parent" | '/a/A.class'
+    "parent" | '/a/b/B.class'
+    "parent" | '/a/b/c/C.class'
+    "child"  | '/x/X.class'
+    "child"  | '/x/y/Y.class'
+    "child"  | '/x/y/z/Z.class'
+  }
+
+
+  def "test read class twice"() {
```
Comment (iNikem) [inline agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy:63]: Don't understand, what does this test do exactly?

File: agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    InternalJarURLHandler handler = new InternalJarURLHandler(dir, testJarLocation)
+    when:
+    handler.openConnection(new File(file).toURI().toURL())
+    then:
+    // not going to specify (and thereby constrain) the sub type because it doesn't matter
+    thrown IOException
+
+    // permuted
+    where:
+    dir      | file
```
Comment (iNikem) [inline agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy:102]: Why do you want to test several non-existent paths here? Why 1 is not enough?

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/matcher/AdditionalLibraryIgnoresMatcher.java [hunk 9]
```diff
@@ -116,7 +116,11 @@ public boolean matches(final T target) {
             || name.startsWith(
                 "org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedServletContainer$")
             || name.equals(
-                "org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedWebappClassLoader")) {
+                "org.springframework.boot.context.embedded.tomcat.TomcatEmbeddedWebappClassLoader")
```
Comment (iNikem) [inline agent-tooling/src/main/java/io/opentelemetry/auto/tooling/matcher/AdditionalLibraryIgnoresMatcher.java:119]: I would love to see some explanations, why these class loaders in particular are listed.

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/muzzle/ReferenceMatcher.java [hunk 10]
```diff
@@ -293,4 +300,19 @@ private static boolean matchesPrimitive(String longName, String shortName) {
     }
     return null;
   }
+
+  private static List<Mismatch> lazyAdd(List<Mismatch> mismatches, Mismatch mismatch) {
```
Comment (iNikem) [inline agent-tooling/src/main/java/io/opentelemetry/auto/tooling/muzzle/ReferenceMatcher.java:304]: What is the motivation of this optimisation?

File: .githooks/pre-commit [hunk 11]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+# http://redsymbol.net/articles/unofficial-bash-strict-mode/
+set -euo pipefail
+IFS=$'\n\t'
+
+if ! ./gradlew spotlessCheck; then
```
Comment (iNikem) [inline .githooks/pre-commit:6]: By committing this file you leave a developer without any way to configure his own pre-commit hooks. I think this is a very bad idea.

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/AgentClassLoader.java [hunk 12]
```diff
@@ -36,11 +36,14 @@
 
   private static final String AGENT_INITIALIZER_JAR = System.getProperty("ota.initializer.jar", "");
 
+  protected final InternalJarURLHandler internalJarURLHandler;
```
Comment (iNikem) [inline agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/AgentClassLoader.java:39]: Why extract this to a field? It is still used only inside one method, no?

File: gradle.properties [hunk 13]
```diff
@@ -1 +1,4 @@
...
+org.gradle.caching=true
+
 org.gradle.jvmargs=-XX:MaxMetaspaceSize=1g
```
Comment (iNikem) [inline gradle.properties:4]: btw, why do we have/need this?

File: gradle/spotless.gradle [hunk 14]
```diff
@@ -2,16 +2,36 @@ apply plugin: 'com.diffplug.gradle.spotless'
...
   }
   groovy {
     licenseHeaderFile rootProject.file('gradle/enforcement/spotless.license.java'), '(package|import|class)'
   }
   scala {
+    scalafmt()
     licenseHeaderFile rootProject.file('gradle/enforcement/spotless.license.java'), '(package|import|public)'
   }
   kotlin {
+    // ktfmt() // Requires newer version of java
```
Comment (iNikem) [inline gradle/spotless.gradle:17]: which one?

File: gradle/spotless.gradle [hunk 15]
```diff
@@ -2,16 +2,36 @@ apply plugin: 'com.diffplug.gradle.spotless'
...
+    ktlint().userData(['indent_size': '2', 'continuation_indent_size': '2'])
     licenseHeaderFile rootProject.file('gradle/enforcement/spotless.license.java'), '(package|import|public)'
   }
+  groovyGradle {
+    greclipse().configFile(rootProject.rootDir.path + '/gradle/enforcement/spotless-groovy.properties')
+  }
+  kotlinGradle {
+    ktfmt()
```
Comment (iNikem) [inline gradle/spotless.gradle:25]: Does not work above but works here?

File: instrumentation/jdbc/jdbc.gradle [hunk 16]
```diff
@@ -21,9 +21,6 @@ javacc {
   }
 }
 
-tasks.withType(com.github.sherter.googlejavaformatgradleplugin.VerifyGoogleJavaFormat).configureEach {
```
Comment (iNikem) [inline instrumentation/jdbc/jdbc.gradle:24]: These are generated files and should be excluded from format check.

File: instrumentation/kafka-clients-0.11/src/main/java/io/opentelemetry/auto/instrumentation/kafkaclients/KafkaProducerInstrumentation.java [hunk 17]
```diff
@@ -93,6 +93,11 @@ public static SpanWithScope onEnter(
 
       callback = new ProducerCallback(callback, span);
 
+      boolean isTombstone = record.value() == null && !record.headers().iterator().hasNext();
+      if (isTombstone) {
+        span.setAttribute("tombstone", true);
```
Comment (iNikem) [inline instrumentation/kafka-clients-0.11/src/main/java/io/opentelemetry/auto/instrumentation/kafkaclients/KafkaProducerInstrumentation.java:98]: This attribute is certainly not from semantic convention. Why should we want this into Otel?

File: instrumentation/spring-webmvc-3.1/spring-webmvc-3.1.gradle [hunk 18]
```diff
@@ -5,7 +5,11 @@ muzzle {
     group = 'org.springframework'
     module = 'spring-webmvc'
     versions = "[3.1.0.RELEASE,]"
-    skipVersions += ['1.2.1', '1.2.2', '1.2.3', '1.2.4'] // broken releases... missing dependencies
+    skipVersions += [
```
Comment (iNikem) [inline instrumentation/spring-webmvc-3.1/spring-webmvc-3.1.gradle:8]: Not sure new format is better :)

File: instrumentation/spring-webmvc-3.1/src/main/java/io/opentelemetry/auto/instrumentation/springwebmvc/WebApplicationContextInstrumentation.java [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+import io.opentelemetry.auto.tooling.Instrumenter;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.beans.factory.support.BeanDefinitionRegistry;
+
+@AutoService(Instrumenter.class)
```
Comment (iNikem) [inline instrumentation/spring-webmvc-3.1/src/main/java/io/opentelemetry/auto/instrumentation/springwebmvc/WebApplicationContextInstrumentation.java:37]: This will benefit from documentation. This is a good place to document how different instrumentation coordinate with each other and for what purpose.

File: java-agent/java-agent.gradle [hunk 20]
```diff
@@ -22,14 +22,14 @@ jar {
...
-    )
+      )
   }
 }
 
 CopySpec isolateSpec(Collection<Task> sourceTasks) {
   return copySpec {
     from(sourceTasks.collect { zipTree(it.archiveFile) }) {
-      into 'auto-tooling-and-instrumentation.isolated'
+      into 'inst'
```
Comment (iNikem) [inline java-agent/java-agent.gradle:32]: Why this change?

File: benchmark-integration/jetty-perftest/jetty-perftest.gradle (hunk 3 above)
Comment (tylerbenson) [reply benchmark-integration/jetty-perftest/jetty-perftest.gradle:13]: unfortunate side effect of the new formatting.  If you know of better settings to make it better align with how intellij does it, please let me know.

File: instrumentation/spring-webmvc-3.1/spring-webmvc-3.1.gradle (hunk 18 above)
Comment (tylerbenson) [reply instrumentation/spring-webmvc-3.1/spring-webmvc-3.1.gradle:8]: yeah, I agree... Intellij does a better job at formatting these, but unfortunately doesn't have a command line tool, so no validation of proper formatting.  If you have a better configuration please suggest.

File: java-agent/java-agent.gradle (hunk 20 above)
Comment (tylerbenson) [reply java-agent/java-agent.gradle:32]: I suggest you review the original PR.

Comment (iNikem) [reply java-agent/java-agent.gradle:32]: Link? :)

File: .githooks/pre-commit (hunk 11 above)
Comment (tylerbenson) [reply .githooks/pre-commit:6]: how do you suggest providing this kind of functionality then?  I feel if we don't provide it, 99% of people won't automatically run it before commit, get a failure in CI, have to recommit due to failures.

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/muzzle/ReferenceMatcher.java (hunk 10 above)
Comment (tylerbenson) [reply agent-tooling/src/main/java/io/opentelemetry/auto/tooling/muzzle/ReferenceMatcher.java:304]: I suggest you review the original PR.

File: .githooks/pre-commit (hunk 11 above)
Comment (iNikem) [reply .githooks/pre-commit:6]: We currently provide an example script and a suggestion in CONTRIBUTING to symlink it here. I think this is enough. If a developer frequently have to recommit due to formatting issues, they will remember to actually use this hook.

I don't think we have to force all developers to use exactly one pre-commit hook blessed by us. Especially if this is "for their benefit".

File: instrumentation/spring-webmvc-3.1/spring-webmvc-3.1.gradle (hunk 18 above)
Comment (anuraaga) [reply instrumentation/spring-webmvc-3.1/spring-webmvc-3.1.gradle:8]: I would lean towards not auto formatting Gradle files if the formatter doesn't give a result we're happy with, they're a very small part of this repo, compared to Java.

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/muzzle/ReferenceMatcher.java (hunk 10 above)
Comment (trask) [reply agent-tooling/src/main/java/io/opentelemetry/auto/tooling/muzzle/ReferenceMatcher.java:305]: added comment

File: gradle.properties (hunk 13 above)
Comment (trask) [reply gradle.properties:4]: good question, created #590 to remove

File: instrumentation/jdbc/jdbc.gradle (hunk 16 above)
Comment (trask) [reply instrumentation/jdbc/jdbc.gradle:24]: spotless appears to be smarter and excludes files under `build/generated`, so this isn't needed anymore

File: java-agent/java-agent.gradle (hunk 20 above)
Comment (trask) [reply java-agent/java-agent.gradle:33]: i added comment

also, the original PRs are all linked in the corresponding commit messages

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java (hunk 1 above)
Comment (trask) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java]: thx! removed

File: benchmark-integration/jetty-perftest/jetty-perftest.gradle (hunk 3 above)
Comment (trask) [reply benchmark-integration/jetty-perftest/jetty-perftest.gradle]: reverted auto-formatting of gradle files for now

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java (hunk 5 above)
Comment (trask) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java]: updated comment

File: agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy (hunk 8 above)
Comment (trask) [reply agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy]: updated

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/AgentClassLoader.java (hunk 12 above)
Comment (trask) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/AgentClassLoader.java]: updated

File: gradle/spotless.gradle (hunk 14 above)
Comment (trask) [reply gradle/spotless.gradle]: is requires Java 11, which is fine for us, so i tried it, but it only supports 4-space indent, i updated comment

File: gradle/spotless.gradle (hunk 15 above)
Comment (trask) [reply gradle/spotless.gradle]: removed gradle auto-formatting now

File: instrumentation/spring-webmvc-3.1/spring-webmvc-3.1.gradle (hunk 18 above)
Comment (trask) [reply instrumentation/spring-webmvc-3.1/spring-webmvc-3.1.gradle]: reverted auto-formatting of gradle files for now

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java (hunk 2 above)
Comment (anuraaga) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java:82]: I sort of lean towards not using the threadlocal optimization for these small strings and let it all be garbage that can be collected, over the complexity of weak references, since I suspect this is a tiny part of the optimization in the attached PR.

But if we're only leaking a few hundred bytes it's not so bad either.

File: instrumentation/kafka-clients-0.11/src/main/java/io/opentelemetry/auto/instrumentation/kafkaclients/KafkaProducerInstrumentation.java (hunk 17 above)
Comment (trask) [reply instrumentation/kafka-clients-0.11/src/main/java/io/opentelemetry/auto/instrumentation/kafkaclients/KafkaProducerInstrumentation.java:98]: we (currently at least) capture lots of span attributes that are not defined by semantic conventions

maybe better if we prefixed this, e.g. `kafka.tombstone`, so it doesn't look like a semantic attribute?

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java (hunk 2 above)
Comment (richardstartin) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java:82]: @trask actually this was an insurance policy against there ever being more than _one_ thread accessing the variable. We haven't seen any negative consequences in our benchmarks, but I think it's a question of "take it or leave it" - if it doesn't look right to you, please just change it.

File: agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy (hunk 7 above)
Comment (iNikem) [reply agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy:48]: Still don't understand :( Why we should specifically test that we can read class on our second attempt? How does second attempt differ from the first or the third one?

File: .githooks/pre-commit (hunk 11 above)
Comment (iNikem) [reply .githooks/pre-commit:6]: A. I wonder, how many developers will be confused by this name similarity. IMO from this POV old location, `buildscripts` was less confusing. But let it be.

File: instrumentation/kafka-clients-0.11/src/main/java/io/opentelemetry/auto/instrumentation/kafkaclients/KafkaProducerInstrumentation.java (hunk 17 above)
Comment (iNikem) [reply instrumentation/kafka-clients-0.11/src/main/java/io/opentelemetry/auto/instrumentation/kafkaclients/KafkaProducerInstrumentation.java:98]: "We are already doing strange things, let us do more strange things" is not exactly good reason :) We already have unanswered questions "why do we have this attribute here", such as `dispatcher.target` or `servlet.origin`. I am hesitant to add more.

File: agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy (hunk 7 above)
Comment (richardstartin) [reply agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy:48]: @iNikem I'm going to assume that you are keeping an open mind here so I will try to explain :).  

There are two reasons related to a caching mechanism:

1. **Constrain what is cached:** I wanted to test that a load does not cache the stream after releasing it and losing control over its lifecyle. What's special about the second load which makes it distinct from the first load is that _there was a **predecessor**_.  There is a caching mechanism in place, (the `JarEntry` is cached with the class name), but what if someone comes along later and tried to cache something closeable like the stream instead? That would be unfortunate, but this test would break if that ever happened.
2. **Hitting the branch where caching has occurred**: there happen to be about 800 classloads early on which go all the way down the chain and end up in the leaf-level classloader _twice_. Preventing this from happening would have been a much more invasive change than mitigating it with a cache. Unless you test the load twice, you won't test the branch where the cache is hit. The logic actually dismisses the cache after a hit, based on knowledge of this pattern, making the third load indistinguishable from the first load, hence the lack of a test for it. Good question though.

Does that make sense? :)

Comment (iNikem) [reply agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy:48]: Great, this explanation does make sense :) What is now missing, is this (or abridged) explanation in the test itself :) For that poor developer who will try to understand this test half a year in the future.

Comment (trask) [reply agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy:48]: thx @richardstartin! @iNikem i will add this explanation to the test 👍

File: instrumentation/kafka-clients-0.11/src/main/java/io/opentelemetry/auto/instrumentation/kafkaclients/KafkaProducerInstrumentation.java (hunk 17 above)
Comment (trask) [reply instrumentation/kafka-clients-0.11/src/main/java/io/opentelemetry/auto/instrumentation/kafkaclients/KafkaProducerInstrumentation.java:98]: i created #602 to track and make decision about this

File: agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy (hunk 7 above)
Comment (richardstartin) [reply agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy:48]: @iNikem I would love to live in a world where each and every test I'm unfamiliar with happens to have a nice two paragraph explanation attached and I never need to go and take a look at the system under test, but it's not the kind of luxury I've experienced myself.

Comment (iNikem) [reply agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy:48]: If our world is not ideal, it does not mean we should not try to make it a better place. I hope you are not saying “There are a lot of bad code out there, why bother writing good one?” :)

Comment (trask) [reply agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy:48]: hm, i don't think that's what @richardstartin was saying. but anyways, i think we're good here. i will add the comment to this test.

Comment (trask) [reply agent-bootstrap/src/test/groovy/io/opentelemetry/auto/bootstrap/InternalJarURLHandlerTest.groovy:63]: comment added 👍

File: agent-tooling/src/main/java/io/opentelemetry/auto/tooling/matcher/AdditionalLibraryIgnoresMatcher.java (hunk 9 above)
Comment (trask) [reply agent-tooling/src/main/java/io/opentelemetry/auto/tooling/matcher/AdditionalLibraryIgnoresMatcher.java:119]: this was not changed in this PR, but class loaders are listed in here so that `java-classloader` instrumentation will not be skipped. the way these are caught and added here is that `AgentTestRunner` does not use `AdditionalLibraryIgnoresMatcher`, and then it validates that there are no classes that were instrumented that would have been skipped by `AdditionalLibraryIgnoresMatcher`, and so the test fails, and then you add suppress the class from being skipped in here

File: agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java (hunk 2 above)
Comment (trask) [reply agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/InternalJarURLHandler.java]: removed the `ThreadLocal<StringBuilder>`
